api/src/utils/xlsx-utils/worksheet-utils.ts

```diff
@@ -499,27 +509,39 @@ export async function getCBMeasurementsFromWorksheet(
   critterBaseService: CritterbaseService,
   sheet = DEFAULT_XLSX_SHEET_NAME
 ): Promise<TsnMeasurementMap> {
-  const tsnMeasurements: TsnMeasurementMap = {};
   const rows = getWorksheetRowObjects(xlsxWorksheets[sheet]);
+  const tsns = rows.map((row) => String(row['ITIS_TSN'] ?? row['TSN'] ?? row['TAXON'] ?? row['SPECIES']));
+  return getCBMeasurementsFromTSN(tsns, critterBaseService);
+}
 
+/**
+ * Fetches measurement definitions from critterbase for a given list of TSNs and creates and returns a map with all data fetched
+ * Throws if a TSN does not return measurements.
+ *
+ * @param {string[]} tsns List of TSNs
+ * @param {CritterbaseService} critterBaseService Critterbase service
+ * @returns {*} Promise<TsnMeasurementMap>
+ */
+export async function getCBMeasurementsFromTSN(
+  tsns: string[],
+  critterBaseService: CritterbaseService
+): Promise<TsnMeasurementMap> {
+  const tsnMeasurements: TsnMeasurementMap = {};
   try {
-    for (const row of rows) {
-      const tsn = String(row['ITIS_TSN'] ?? row['TSN'] ?? row['TAXON'] ?? row['SPECIES']);
+    for (const tsn of tsns) {
       if (!tsnMeasurements[tsn]) {
-        // TODO: modify Critter Base to accept multiple TSN at once
         const measurements = await critterBaseService.getTaxonMeasurements(tsn);
         if (!measurements) {
-          // TODO: Any data for a TSN that has no measurements is invalid and should reject the uploaded CSV
+          throw `No measurements found for tsn: ${tsn}`;
         }
 
         tsnMeasurements[tsn] = measurements;
       }
     }
   } catch (error) {
```

Comment:
Assuming you accept the above comment, you can type anotate this, and refactor the throw on Line 543:
```javascript
} catch (error: Error) {
  throw new ApiGeneralError(`Error connecting to the Critterbase API: ${error.message}`);
}
```

Reply:
the catch error cannot be typed like that, the error value is `any` or `unknown` and there is no guarantee of that the thing thrown is of `Error` type. In a catch like that you'd need do a check like 
`if (error instanceof Error)` to enforce a type on error

Reply:
Currently it looks like we're just including the entire `error` object in the `ApiGeneralError` constructor, which is fine by me. Feel free to ignore the above comment

Reply:
Never fully built out the custom error classes, though I think in an ideal world, we would always throw some kind of custom known error, just to ensure consistency, and making it more obvious to upstream code what errors could be expected.